### DISCUSSION TITLE
Lazy load plugins using lz.n

### DISCUSSION THREAD

**Discussion by arunoruto** [nix-community/nixvim#1822] [Q&A]
I am following issue #421, which covers possible solutions for lazy-loading plugins in `nixvim` so that `neovim` can become a bit snappier. Recently, @GaetanLepage recommended [lz.n](https://github.com/nvim-neorocks/lz.n), which is a sleek solution to the problem. It only does lazy loading, but supposedly, it is doing it well! And if `nixvim` is used, we do not need the plugin manager aspect of `lazy.vim`.

I wanted to see if someone is using it so we can share our config and experiences! My current config of `lz.n` is [here](https://github.com/arunoruto/flake/blob/0db4d0edaafde5e67ab5908ff508729c18770019/home-manager/server/shell/nvim/plugins/lz-n.nix). I tried to load `nvim-cmp` only when insert mode is entered [here](https://github.com/arunoruto/flake/blob/0db4d0edaafde5e67ab5908ff508729c18770019/home-manager/server/shell/nvim/plugins/coding/completion/cmp.nix#L87). Somehow, I am not sure I am doing it right, since [vim-startuptime](https://github.com/rhysd/vim-startuptime) still shows `nvim-cmp` and other plugins I tried lazy-loading in the resulting profile, but should be only loaded, if a certain command is executed!

**Comment by MattSturgeon** [comment]
I haven't looked for a while, but I think that version of lazy loading doesn't override the runtimepath (like lazy.nvim does).

Instead it just focuses on _not_ calling `setup` until necessary.

If my above understanding is correct, this means you will see all plugins as loaded, they just won't have been "set up" yet.

I'm not aware of any examples using nixvim and lz.n, but I think a good first step would be to have a nixvim module similar to our lazy.nvim and packer modules.

**Reply by arunoruto** [reply]
> I haven't looked for a while, but I think that version of lazy loading doesn't override the runtimepath (like lazy.nvim does).
> 
> Instead it just focuses on _not_ calling `setup` until necessary.
> 
> If my above understanding is correct, this means you will see all plugins as loaded, they just won't have been "set up" yet.
> 
Thanks for the clarification! I thought loading and setting up a plugin were tied together, but it's a nice thing they can be split! But if I am correct, I have not seen a performance difference (startup time), so I thought I did something wrong...
> I'm not aware of any examples using nixvim and lz.n, but I think a good first step would be to have a nixvim module similar to our lazy.nvim and packer modules.
>
I guess `lz.n` would first have to be awalable in `nixpkgs` under `vimPlugins`? I have never packaged something for nixpkgs, but maybe this would be a good first try :)
